Classify the email using the binary classification method. Email Spam detection has two
states:
a) Normal State – Not Spam, 
b) Abnormal State – Spam. 
Use K-Nearest Neighbors and
Support Vector Machine for classification. Analyze their performance.

In [27]:
import pandas as pd
import numpy as np

In [28]:
df = pd.read_csv("emails.csv")
df.head()

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

In [29]:
df.isnull().sum()    

Email No.     0
the           0
to            0
ect           0
and           0
             ..
military      0
allowing      0
ff            0
dry           0
Prediction    0
Length: 3002, dtype: int64

In [30]:
df.drop(columns=["Email No."], inplace=True)
X = df.drop(columns=["Prediction"])
y = df["Prediction"]

split dataset

In [33]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Standardize features

In [34]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

 Train and Test K-Nearest Neighbors (KNN) Model

In [35]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

# Predictions
y_pred_knn = knn.predict(X_test)

# Evaluation
print("K-Nearest Neighbors (KNN) Performance")
print("Accuracy:", accuracy_score(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))
print(confusion_matrix(y_test, y_pred_knn))


K-Nearest Neighbors (KNN) Performance
Accuracy: 0.8253865979381443
              precision    recall  f1-score   support

           0       0.98      0.77      0.86      1097
           1       0.63      0.96      0.76       455

    accuracy                           0.83      1552
   macro avg       0.81      0.86      0.81      1552
weighted avg       0.88      0.83      0.83      1552

[[846 251]
 [ 20 435]]


Train and Test Support Vector Machine (SVM) Model

In [36]:
from sklearn.svm import SVC
svm = SVC(kernel="linear")
svm.fit(X_train, y_train)

# Predictions
y_pred_svm = svm.predict(X_test)

# Evaluation
print("Support Vector Machine (SVM) Performance")
print("Accuracy:", accuracy_score(y_test, y_pred_svm))
print(classification_report(y_test, y_pred_svm))
print(confusion_matrix(y_test, y_pred_svm))


Support Vector Machine (SVM) Performance
Accuracy: 0.9400773195876289
              precision    recall  f1-score   support

           0       0.96      0.95      0.96      1097
           1       0.89      0.91      0.90       455

    accuracy                           0.94      1552
   macro avg       0.92      0.93      0.93      1552
weighted avg       0.94      0.94      0.94      1552

[[1043   54]
 [  39  416]]


K-Nearest Neighbors (KNN):

Accuracy: 82.5%, meaning about 82.5% of all predictions were correct.
Precision and Recall:

For class 0 (Not Spam): Prec
ision is high at 0.98 but recall is lower at 0.77, showing KNN correctly identified Not Spam emails but missed some, with 251 false positives.

For class 1 (Spam): Precision is 0.63 and recall is high at 0.96, meaning KNN flagged most actual spam emails, though some were incorrectly labeled as spam.

Confusion Matrix: There were 846 true negatives, 251 false positives, 20 false negatives, and 435 true positives.
Support Vector Machine (SVM):

Accuracy: 94%, indicating higher overall performance than KNN.
Precision and Recall:

Class 0: High precision (0.96) and recall (0.95), so SVM more accurately identified Not Spam emails.

Class 1: Precision (0.89) and recall (0.91) are balanced, indicating SVM identified most spam emails correctly.

Confusion Matrix: With 1043 true negatives, 54 false positives, 39 false negatives, and 416 true positives, SVM outperformed KNN with fewer misclassifications overall.

In summary, SVM was more accurate and balanced in performance, especially with better precision and recall for both classes, making it a stronger choice for email spam detection in this dataset.